### ISSUE TITLE
Regarding Jakarta EE 10 implementation of BIRT to work with wildfly 27

### ISSUE THREAD

**Issue by kranthirathikanti** [eclipse-birt/birt#1172]
Hi Team,

We have been using the eclipse BIRT in our product and we are planning to support wildfly 27 to our product and which is a  Jakarta EE 10 implementation. As part of this support, BIRT also need to be with Jakarta EE 10 implementation. It seems that the latest version of eclipse BRT (4.12.0) doesn't have the Jakarta EE 10 implementation. Could you please let us know if you have a plan for this implementation ? if so, any timeline.

**Comment by brivalin** [comment]
Hello there,

To add some additional informations, what I saw not working on Jakarta EE 10 at this moment is the **Viewer Tag Library**.

When trying to deploy a project including the "viewerservlets.jar" with tag library enabled, it runs into lots of "java.lang.NoClassDefFoundError" exceptions.
By example the first one looks like :

> WARN  [org.jboss.modules.define] (MSC service thread 1-1) Failed to define class org.eclipse.birt.report.taglib.AbstractBaseTag in Module "xxx.war" from Service Module Loader: java.lang.NoClassDefFoundError: Failed to link org/eclipse/birt/report/taglib/AbstractBaseTag (Module "xxx.war" from Service Module Loader): javax/servlet/jsp/tagext/TagSupport

And finally, the deployment ends with : 

> WARN  [org.jboss.as.ee] (MSC service thread 1-1) WFLYEE0007: Not installing optional component org.eclipse.birt.report.taglib.ParamDefTag due to an exception

The new Jakarta EE 10 specifications replaced all "javax" package namespaces by the "jakarta" one that I think causes all of these  java.lang.NoClassDefFoundError exceptions.

**Comment by kranthirathikanti** [comment]
Thank you for your time and response.

Yes, you are right.  The new Jakarta EE 10 specification expects a jakarta package namespace instead of "javax".

So, viewerservlets.jar jar should implement the new specification in order to support  Jakarta EE 10.

Thanks,
Kranthi

**Comment by wimjongman** [comment]
Thanks for the analysis. Can you provide a patch? The [Contribution Guide](https://github.com/eclipse/birt/blob/master/CONTRIBUTING.md) is here.

**Comment by dzorwulu** [comment]
Hello Team
Please what is the status of this issue? In BIRT-runtime-4.13.0  classes in viewservlets.jar still have 'javax' imports instead or 'jakarta'. This is breaking our app server with java 17.

**Comment by wimjongman** [comment]
We have not worked on it. Please, members of this discussion, provide a patch. The [Contribution Guide](https://github.com/eclipse/birt/blob/master/CONTRIBUTING.md) is here.

**Comment by claesrosell** [comment]
I would just like to note that we currently are using Jetty10, which some parts of the Eclipse 2022-12 release rely on. A move to the jakarta namespace would require us to use Jetty11 instead.

I have, in principle, nothing against this, but it would probably require us to maintain two separate branches of BIRT since the javax namespace is far from dead.

Some of the tools we are using, like axis 1.4, rely on the javax namespace. We probably would need to move away from those.

**Comment by dzorwulu** [comment]
Thanks for the prompt response
'javax' name is proprietary and belongs to Oracle. 'jakarta' is the replacement open-source name for all 'javax' named packages.   

Oracle JDK road & LTS
https://www.oracle.com/java/technologies/java-se-support-roadmap.html#:~:text=Support%20for%20JavaFX%20on%20Java%20SE%208%20will%20continue%20until%20March%202025.&text=*%20Oracle%20Java%20SE%20product%20dates,up%2Dto%2Ddate%20information.

Axis2 1.8
https://axis.apache.org/axis2/java/core/release-notes/1.8.0.html

 regards

**Comment by speckyspooky** [comment]
@wimjongman , @claesrosell , @merks 

Short note: the report viewer isn't moved currently to "jakarta".
The current build re-mastering will be done with JDK17.

Please correct my interpretation, in my understanding the viewer should be moved
to "jakarta" before we publish the new BIRT-release 4.14 under JDK17.

**Comment by merks** [comment]
I imagine moving to Jakarta might be a significantly time consuming process and it's not at all clear that all your libraries/dependencies can even accommodate that!   Furthermore, I don't see that building with Java 17 is in any way related to Jakarta.  The JDK does not contain, use, nor provide anything from the jakarta namespace nor will it ever do so, so there is no relation.

**Comment by speckyspooky** [comment]
Sounds good for me. In my mind the usage of JDK17 is directly combined with the module handling and the removed javax from java-core.
So it seems to be build with the existing Java-references.

**Comment by merks** [comment]
No, modules kicked in for Java 9 and only parts of javax were migrated to jakarta, i.e., only the JEE things.  So not to worry of course eventually one will want to migrate.  But I would plan for such a thing at the start of a release cycle.  Even the Eclipse Platform itself is only in early stages of such migration...

**Comment by dm-ion** [comment]
Hi, it is not exactly "clean" but for your information we are running BIRT within a Jakarta project on tomcat 10.x without issue. We use the BIRT engine and BIRT web viewer within our web application to run a variety of different reports. We certainly do not use all of the features that BIRT exposes, so this may not work "everywhere" or for "everything".

We accomplished this by using the tomcat translator to automatically translate all of the BIRT jars to Jakarta. In terms of third parties it was mostly okay, but we did have to also translate axis and make one api change there for Servlet 6.

There are some other peculiarities to our build as well, such as not using OSGI, subbing in maven central artifacts for the orbit ones where we can, some other dependency upgrades, etc. In addition the tomcat translator does other undesirable things to your jar files, such as stripping signatures.

Certainly would be some work to convert BIRT to Jakarta upstream. Simply commenting for anyone who needs to run BIRT on Jakarta sooner, it's certainly possible.

**Comment by ipanera** [comment]
> Hi, it is not exactly "clean" but for your information we are running BIRT within a Jakarta project on tomcat 10.x without issue. We use the BIRT engine and BIRT web viewer within our web application to run a variety of different reports. We certainly do not use all of the features that BIRT exposes, so this may not work "everywhere" or for "everything".
> 
> We accomplished this by using the tomcat translator to automatically translate all of the BIRT jars to Jakarta. In terms of third parties it was mostly okay, but we did have to also translate axis and make one api change there for Servlet 6.
> 
> There are some other peculiarities to our build as well, such as not using OSGI, subbing in maven central artifacts for the orbit ones where we can, some other dependency upgrades, etc. In addition the tomcat translator does other undesirable things to your jar files, such as stripping signatures.
> 
> Certainly would be some work to convert BIRT to Jakarta upstream. Simply commenting for anyone who needs to run BIRT on Jakarta sooner, it's certainly possible.

Hi @dm-ion.
They are goog news that we are glad to ear because we have the same issue.
It could be nice if you could share more about your changes and pom files with us.
Greetings

**Comment by ahsin019** [comment]
We are trying to migrate it to use EE10, but found out ViewerServlet/ViewerServletContextListener in 4.14.0 still use

javax.servlet.ServletConfig;
javax.servlet.ServletException;
javax.servlet.http.HttpServletRequest;
javax.servlet.http.HttpServletResponse;
javax.servlet.ServletContextEvent
javax.servlet.ServletContextListener

https://github.com/eclipse-birt/birt/blob/master/viewer/org.eclipse.birt.report.viewer/birt/WEB-INF/classes/org/eclipse/birt/report/servlet/ViewerServlet.java


Any plan to to replace above with jakarta?

**Comment by speckyspooky** [comment]
Yes, the replacement is planned, but won't be upcoming directly with the next release
because there are lot of more dependencies which we would have to change if we switch to Jakarta EE.

**Comment by ahsin019** [comment]
Thanks @speckyspooky

Any time frame we can expect BIRT to be compatible with Jakarta EE 10?  Especially since everybody is forcing "java 17" aka.. spring etc.

**Comment by wimjongman** [comment]
Hey Roel, we prioritize it, but we are all unpaid volunteers. If Upsilon would like to sponsor this, please contact me. I can make a quote and work with the BIRT developers to make this happen.

**Comment by speckyspooky** [comment]
Latest update on it BIRT 4.14 is working with WildFly 30.0.1

**Comment by cusmelee** [comment]
I found this on youtube, just for reference: https://www.youtube.com/watch?v=bO3_pjuSjYQ
How to Easily Upgrade to Jakarta EE 10 Javax to Jakarta Prefix without a Hassle

**Comment by Girish-tech1** [comment]
Has anyone tried using https://tomcat.apache.org/download-migration.cgi to migrate BIRT code to work with Jakarta EE?
Is this recommended by BIRT team

**Comment by saumya2907** [comment]
I still see javax namespace in the coreapi  4.16 version 
https://github.com/eclipse-birt/birt/blob/master/core/org.eclipse.birt.core/src/org/eclipse/birt/core/framework/PlatformServletContext.java

is this going to go soon?

**Comment by Jokerinout** [comment]
> Hi, it is not exactly "clean" but for your information we are running BIRT within a Jakarta project on tomcat 10.x without issue. We use the BIRT engine and BIRT web viewer within our web application to run a variety of different reports. We certainly do not use all of the features that BIRT exposes, so this may not work "everywhere" or for "everything".
> 
> We accomplished this by using the tomcat translator to automatically translate all of the BIRT jars to Jakarta. In terms of third parties it was mostly okay, but we did have to also translate axis and make one api change there for Servlet 6.
> 
> There are some other peculiarities to our build as well, such as not using OSGI, subbing in maven central artifacts for the orbit ones where we can, some other dependency upgrades, etc. In addition the tomcat translator does other undesirable things to your jar files, such as stripping signatures.
> 
> Certainly would be some work to convert BIRT to Jakarta upstream. Simply commenting for anyone who needs to run BIRT on Jakarta sooner, it's certainly possible.

Hi, for information we succesfully make Birt working with tomcat10 (Jakarta EE 9) only using the tomcat translator to automatically translate all of the BIRT jars to Jakarta, this one : https://tomcat.apache.org/download-migration.cgi

**Comment by cusmelee** [comment]
@Jokerinout  does it work with BIRT Viewer?

**Comment by Jokerinout** [comment]
> @Jokerinout does it work with BIRT Viewer?

I haven't tried it

**Comment by MilosHavranek** [comment]
If the migration is not planned, jakarta transformed runtime packages could be provided in the meantime.

**Comment by merks** [comment]
More work, more things to test, and more things to publish doesn't sound all that appealing.   😞   

Contribution makes anything possible.  👼

**Comment by cusmelee** [comment]
Let's vote for this the TOP priority!

**Comment by roeltje25** [comment]
> Unlike normal elections, each voter should assume that a vote costs 1000€ with the added benefit that you can vote as many times as you wish! 🥳 I'm sure we'd quickly have the funding to enable all the necessary work. 😈

Like I said in the ticket 2275:

I understand that it's all free time work. It's just that it will make companies like ours to move away from BIRT. That's not a criticism it's just reality. Sorry. 

still kudos for all the hard work

**Comment by ahsin019** [comment]
It looks like ViewerServlet is still using javax.servlet instead of jakarta.servlet ?   
4.22 in repository is compiled with JDK21, any backward compatible like JDK17? 


package org.eclipse.birt.report.servlet;

import java.io.IOException;
import javax.servlet.ServletConfig;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.eclipse.birt.core.exception.BirtException;
import org.eclipse.birt.report.context.BirtContext;
import org.eclipse.birt.report.context.IContext;
import org.eclipse.birt.report.presentation.aggregation.IFragment;
import org.eclipse.birt.report.presentation.aggregation.layout.FramesetFragment;
import org.eclipse.birt.report.presentation.aggregation.layout.RunFragment;
import org.eclipse.birt.report.service.BirtReportServiceFactory;
import org.eclipse.birt.report.service.BirtViewerReportService;
import org.eclipse.birt.report.service.api.InputOptions;
import org.eclipse.birt.report.utility.BirtUtility;

public class ViewerServlet extends BirtSoapMessageDispatcherServlet {
    private static final long serialVersionUID = 1L;

    protected void __init(ServletConfig config) {
        BirtReportServiceFactory.init(new BirtViewerReportService(config.getServletContext()));
        this.viewer = new FramesetFragment();
        this.viewer.buildComposite();
        this.viewer.setJSPRootPath("/webcontent/birt");
        this.run = new RunFragment();
        this.run.buildComposite();
        this.run.setJSPRootPath("/webcontent/birt");
    }
